lite2/client.go

```diff
@@ -0,0 +1,419 @@
+// solution with trust tradeoffs is that we establish an https based protocol
+// with a default end point that populates this information. Also an on-chain
+// registry of roots-of-trust (e.g. on the Cosmos Hub) seems likely in the
+// future.
+type TrustOptions struct {
+	// tp: trusting period.
+	//
+	// Should be significantly less than the unbonding period (e.g. unbonding
+	// period = 3 weeks, trusting period = 2 weeks).
+	//
+	// More specifically, trusting period + time needed to check headers + time
+	// needed to report and punish misbehavior should be less than the unbonding
+	// period.
+	Period time.Duration
+
+	// Header's Height and Hash must both be provided to force the trusting of a
+	// particular header.
+	Height int64
+	Hash   []byte
+}
+
+type mode byte
+
+const (
+	sequential mode = iota + 1
+	skipping
+)
+
+// Option sets a parameter for the light client.
+type Option func(*Client)
+
+// SequentialVerification option configures the light client to sequentially
+// check the headers. Note this is much slower than SkippingVerification,
+// albeit more secure.
+func SequentialVerification() Option {
+	return func(c *Client) {
+		c.verificationMode = sequential
+	}
+}
+
+// SkippingVerification option configures the light client to skip headers as
+// long as {trustLevel} of the old validator set signed the new header. The
+// bisection algorithm from the specification is used for finding the minimal
+// "trust path".
+//
+// trustLevel - fraction of the old validator set (in terms of voting power),
+// which must sign the new header in order for us to trust it. NOTE this only
+// applies to non-adjusted headers. For adjusted headers, sequential
+// verification is used.
+func SkippingVerification(trustLevel cmn.Fraction) Option {
+	if err := ValidateTrustLevel(trustLevel); err != nil {
+		panic(err)
+	}
+	return func(c *Client) {
+		c.verificationMode = skipping
+		c.trustLevel = trustLevel
+	}
+}
+
+// AlternativeSources option can be used to supply alternative providers, which
+// will be used for cross-checking the primary provider.
+func AlternativeSources(providers []provider.Provider) Option {
+	return func(c *Client) {
+		c.alternatives = providers
+	}
+}
+
+// Client represents a light client, connected to a single chain, which gets
+// headers from a primary provider, verifies them either sequentially or by
+// skipping some and stores them in a trusted store (usually, a local FS).
+//
+// Default verification: SkippingVerification(DefaultTrustLevel)
+type Client struct {
+	chainID          string
+	trustingPeriod   time.Duration // see TrustOptions.Period
+	verificationMode mode
+	trustLevel       cmn.Fraction
+
+	// Primary provider of new headers.
+	primary provider.Provider
+
+	// Alternative providers for checking the primary for misbehavior by
+	// comparing data.
+	alternatives []provider.Provider
+
+	// Where trusted headers are stored.
+	trustedStore    store.Store
+	trustedHeader   *types.SignedHeader // H
```

Comment:
I believe this is the highest trusted header from the store. Some comments would help.
It is updated as we add more trusted headers, however we don't "expire" it. I assume the error ErrOldHeaderExpired raised in Verify() bubbles up to a caller that know what to do (e.g. restart the client with a new trusted header)

Reply:
Good point re expiring. However, I don't think it makes sense to write about it here since "expired" is not inherent property of the Header. I will add a check and a comment to `TrustedHeader()` method